pilot/pkg/config/mcp/controller.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"istio.io/istio/pilot/pkg/model"
+	mcpclient "istio.io/istio/pkg/mcp/client"
+)
+
+var (
+	errNotFound      = errors.New("item not found")
+	errAlreadyExists = errors.New("item already exists")
+	errUnsupported   = errors.New("this operation is not supported by mcp controller")
+)
+
+//go:generate counterfeiter -o fakes/logger.go --fake-name Logger . logger
+type logger interface {
+	Warnf(template string, args ...interface{})
+	Infof(template string, args ...interface{})
+}
+
+type CoreDataModel interface {
+	model.ConfigStoreCache
+	mcpclient.Updater
+}
+
+type changeEvent struct {
+	metaName       string
+	resource       proto.Message
+	descriptorType string
+}
+
+type Controller struct {
+	configStore              map[string]map[string]*sync.Map
+	eventCh                  chan func(model.Config, model.Event)
+	updateCh                 chan changeEvent
+	descriptorsByMessageName map[string]model.ProtoSchema
+	logger                   logger
+}
+
+func NewController(logger logger) CoreDataModel {
```

Comment:
See above - we don't pass loggers as params, and don't add extra interfaces for loggers. Use the existing logger as used by everyone else.

Reply:
👍